### PR TITLE
Use redis pipeline

### PR DESCRIPTION
**Pull request by Anton-Shutik**
Speeds up `Segment` refreshing _a lot_.
I've tried that on our 90 segments and on that code it completed refreshing in ~1 hour!. 
A segment with 1.4m users in it was refreshed in ~ 3 minutes, while without the pipeline it took ~30 mins
**UPD:**
Added `count=1000` to `scan` command and now it refreshes all segments in ~20 mins (in a single thread), segment with 2.7m ids in ~ 2 mins. That was tried on staging

>A Request/Response server can be implemented so that it is able to process new requests even if the client didn't already read the old responses. This way it is possible to send multiple commands to the server without waiting for the replies at all, and finally read the replies in a single step.

[Redis docs, pipelining](https://redis.io/topics/pipelining)

Risky part is that redis keeps in memory results from all commands you send in pipeline. We use pipeline for `sadd` and `srem` which return `bool` type, so I think we should be good to go with 10k commands per batch.

Probably there was a reason pipelining was not used during initial development, may be @chrisclark or @katekligman knows

### CODE REVIEW

**Comment by sramirezgroveco** [comment]
Yeah, let's see if we can get some answers around why pipelining wasn't used first to understand the implications.

**Comment by Anton-Shutik** [comment]
> Yeah, let's see if we can get some answers around why pipelining wasn't used first to understand the implications.

Yeah, that would be great. I'm not sure we will get it, since initially the app was created by Chris and stored the data in the database, and then, Kate moved it to redis. Probably we just can try it. 

I've already played with that on staging and it works _way_ faster. A segment of ~1.4m members in it refreshed in 30mins vs 3 minutes on this code

**Comment by chrisclark** [comment]
@sramirezgroveco I suspect it just wasn’t considered vs explicitly not used. But that’s a total guess. I can’t think of a reason (ever) to not use pipelining. I’m in favor of this change.

File: segments/helpers.py [hunk 1]
```diff
@@ -136,16 +136,16 @@ def refresh_segment(self, segment_id, sql):
 
     def delete_segment(self, segment_id):
         segment_key = self.segment_key % segment_id
-        for user_id in self.redis.sscan_iter(segment_key):
-            self.remove_segment_membership(segment_id, user_id)
+        cursor = '0'
```
Comment (apolisskaya) [inline segments/helpers.py]: This seems a bit hacky, we can probably find a better way to break out of the while loop

**Comment by katekligman** [comment]
> @sramirezgroveco I suspect it just wasn’t considered vs explicitly not used. But that’s a total guess. I can’t think of a reason (ever) to not use pipelining. I’m in favor of this change.

@chrisclark That's correct.  This was my first substantial Redis project and I wasn't aware of the feature.

@Anton-Shutik Good job on taking the system forward.  Miss you all!

File: segments/helpers.py (hunk 1 above)
Comment (Anton-Shutik) [reply segments/helpers.py]: @apolisskaya That's exactly how they do it in `sscan_iter`

File: segments/helpers.py [hunk 2]
```diff
@@ -112,17 +103,25 @@ def refresh_segment(self, segment_id, sql):
...
             new_key
         )
-        for user_key in (self.segment_member_key % i for i in self.redis.sscan_iter(new_key)):
-            self.redis.sadd(user_key, segment_id)
+        with self.redis.pipeline(transaction=False) as pipeline:
+            for user_id in self.redis.sscan_iter(new_key, count=REDIS_SSCAN_COUNT):
+                pipeline.sadd(self.segment_member_key % user_id, segment_id)
+                if len(pipeline) >= BATCH_SIZE:
+                    pipeline.execute()
+            pipeline.execute()
```
Comment (sramirezgroveco) [inline segments/helpers.py]: Can we move this to a function call? It overlaps with line 119 for the most part other than the pipeline operation. I think you can basically push this and just send the operation to do to the function. Something like

```
run_pipeline(new_key, REDIS_SSCAN_COUNT, "sadd")

def run_pipeline(new_key, count, operation):
  with self.redis.pipeline(transaction=False) as pipeline:
      for user_id in self.redis.sscan_iter(new_key, count=REDIS_SSCAN_COUNT):
          pipeline.operation(self.segment_member_key % user_id, segment_id)
          if len(pipeline) >= BATCH_SIZE:
              pipeline.execute()
      pipeline.execute()
  

```

**Comment by Anton-Shutik** [comment]
@sramirezgroveco done

**Pull request merged**